lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+blockSeqLastUTxO = TxSeq.lastUTxO . blockSeqToTxSeq
+
+-- | Retrieves, from a block sequence, the complete list of transactions that
+--   are expected to be relevant to a particular wallet.
+--
+-- The returned transactions are listed in the same order that they appear
+-- within the block sequence.
+--
+blockSeqOurTxs
+    :: forall s. (IsOurs s Address, IsOurs s RewardAccount)
+    => s
+    -> BlockSeq
+    -> [Tx]
+blockSeqOurTxs s0 blockSeq = blockSeq
+    & blockSeqToTxSeq
+    & TxSeq.toTransitionList
+    & filterM isOurTransitionM
+    & flip evalState s0
+    & fmap (\(_, tx, _) -> tx)
+  where
+    isOurAddressM :: Address -> State s Bool
+    isOurAddressM = fmap isJust . state . isOurs
+
+    isOurTransitionM :: (UTxO, Tx, UTxO) -> State s Bool
+    isOurTransitionM (u, tx, _) =
+        -- Here we use the test function 'isOurTx', which is not used by the
+        -- implementation of 'applyBlocks'.
+        isOurTx tx =<< UTxO.filterByAddressM isOurAddressM u
+
+-- | Extracts a 'TxSeq' from a 'BlockSeq'.
+--
+-- The resulting 'TxSeq' does retain knowledge of block boundaries, but does
+-- not have any knowledge of block heights or slot numbers.
+--
+blockSeqToTxSeq :: BlockSeq -> TxSeq
+blockSeqToTxSeq BlockSeq {shrinkableTxSeq} = getTxSeq shrinkableTxSeq
+
+--------------------------------------------------------------------------------
+-- Convenience functions for calling 'applyBlocks' and processing the result
+--------------------------------------------------------------------------------
+
+-- | Embeds a 'UTxO' set within a 'Wallet' suitable for use with 'applyBlocks'.
+--
+utxoToWallet :: s -> UTxO -> Wallet s
+utxoToWallet s u = unsafeInitWallet u currentTip s
+  where
+    -- The 'currentTip' field is not currently used by 'applyBlocks'.
+    --
+    currentTip :: BlockHeader
+    currentTip = shouldNotEvaluate "currentTip"
+
+    shouldNotEvaluate :: String -> a
+    shouldNotEvaluate name = error $ unwords
+        [name, "was unexpectedly evaluated"]
+
+-- | Extracts a list of filtered transactions from the result of 'applyBlocks'.
+--
+-- The returned transactions are listed in the same order that they appear
+-- within the filtered blocks.
+--
+applyBlocksFilteredTxs
+    :: NonEmpty ([FilteredBlock], (DeltaWallet s, Wallet s)) -> [Tx]
+applyBlocksFilteredTxs = mconcat . mconcat . NE.toList
+    . fmap (fmap (fmap fst . view #transactions) . fst)
+
+-- | Extracts out the last UTxO set returned by 'applyBlocks'.
+--
+-- This UTxO set represents the final state of the wallet after applying
+-- just those transactions that are relevant to the wallet.
+--
+applyBlocksLastUTxO
+    :: NonEmpty ([FilteredBlock], (DeltaWallet s, Wallet s)) -> UTxO
+applyBlocksLastUTxO = utxo . snd . snd . NE.last
+
+-- | Removes a fee from a transaction.
+--
+-- Useful in situations where we want to compare transactions without comparing
+-- their fees.
+--
+nullifyFee :: Tx -> Tx
+nullifyFee = set #fee Nothing
```

Comment:
IMO this function is below the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)

Reply:
> IMO this function is below the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)

Fair enough!

I had originally intended to extend this function with knowledge of how `applyBlocks` modifies fees, but eventually didn't get around to that. After giving the matter some thought, I'd probably be in favour of modifying `applyBlocks` so that it doesn't modify the `fee` field at all. (This behaviour seems completely unnecessary.)

I've inlined the definition and removed the function in 96438d11433aba1b537bfdc1469ebf469b51f7f3.